### PR TITLE
Add F.normalize

### PR DESCRIPTION
**Pull request by szagoruyko**
with gradcheck and doc.

### CODE REVIEW

**Comment by soumith** [comment]
rename lp_normalize to normalize, and remove l2_normalize

**Comment by vadimkantorov** [comment]
Should default dim = -1?

**Comment by szagoruyko** [comment]
@soumith pushed
@vadimkantorov we could then also have p=2, idk

**Comment by soumith** [comment]
yes, please put p=2 as the default (you will have to either have a default dim argument, or move p to 3rd argument)

File: torch/nn/functional.py [hunk 1]
```diff
@@ -742,3 +742,16 @@ def triplet_margin_loss(anchor, positive, negative, margin=1.0, p=2, eps=1e-6, s
...
+
+def normalize(input, p, dim, eps=1e-12):
+    r"""Performs l_p normalization of inputs over specified dimension.
+
+    Does:
+
+    .. math::
+        v = \frac{v}{\max(\lVert v \rVert_p, \epsilon)}
+
+    for each subtensor v over dimension dim of input.
```
Comment (apaszke) [inline torch/nn/functional.py]: it might be worth adding that each `v` is viewed as a vector (if `input` is a 3D tensor, then the sub-matrices **wouldn't** be normalized w.r.t. a matrix norm).

File: torch/nn/functional.py [hunk 2]
```diff
@@ -742,3 +742,16 @@ def triplet_margin_loss(anchor, positive, negative, margin=1.0, p=2, eps=1e-6, s
     dist_hinge = torch.clamp(margin + d_p - d_n, min=0.0)
     loss = torch.mean(dist_hinge)
     return loss
+
+
+def normalize(input, p, dim, eps=1e-12):
+    r"""Performs l_p normalization of inputs over specified dimension.
```
Comment (apaszke) [inline torch/nn/functional.py]: maybe wrap L_p in :math:?

File: torch/nn/functional.py [hunk 3]
```diff
@@ -742,3 +742,16 @@ def triplet_margin_loss(anchor, positive, negative, margin=1.0, p=2, eps=1e-6, s
     dist_hinge = torch.clamp(margin + d_p - d_n, min=0.0)
     loss = torch.mean(dist_hinge)
     return loss
+
+
+def normalize(input, p, dim, eps=1e-12):
```
Comment (apaszke) [inline torch/nn/functional.py]: p should be optional and default to 2.

File: torch/nn/functional.py [hunk 4]
```diff
@@ -744,14 +744,17 @@ def triplet_margin_loss(anchor, positive, negative, margin=1.0, p=2, eps=1e-6, s
...
     Does:
 
     .. math::
         v = \frac{v}{\max(\lVert v \rVert_p, \epsilon)}
 
-    for each subtensor v over dimension dim of input.
+    for each subtensor v over dimension dim of input. Each subtensor is flattened into a vector,
+    i.e. :math:`\lVert v \rVert_p` is not a matrix norm.
+
+    With default arguments normalizes over the last dimension with Euclidean norm.
```
Comment (apaszke) [inline torch/nn/functional.py]: Wouldn't it make more sense to default to normalization over the 1st (feature) dim? Also, can you add the docstrings for each argument?

Comment (vadimkantorov) [inline torch/nn/functional.py]: Sometims you'd have (batchdim x numrois x embedding) layout from Fast-RCNN-like pipelines, dim = -1 is more conventient in such cases...

Comment (apaszke) [inline torch/nn/functional.py]: I'm not saying there are no cases where a different dim is needed, I'm only thinking what is more common. I think that in most applications last dim will be time or width and it seems weird to normalize over that.

Comment (szagoruyko) [inline torch/nn/functional.py]: I'll side with Adam here - we assume NCHW most of the time.

**Review by apaszke — APPROVED** [review]
